doc/debug_framework_design_spec.md

```diff
@@ -0,0 +1,394 @@
+
+**Component registration options**
+
+Framework only manages component registration and invocation of callback functions. Framework itself does not have an executing context. To execute the callback function, components have an option to instruct framework either to create an additional thread that runs in the registrant context or not create a thread but merely publish an event to Redis and expect component to handle the event.  
+Components have an option to register with debug framework using any one of the below APIs:  
+Option #1. `Debugframework::linkWithFramework()`  
+Option #2. `Debugframework::linkWithFrameworkNoThread()`  
+
+Option #1 is preferred. However if a component already handles RedisDB events and doesn't want an additional thread (to avoid thread synchronization issues) might opt for option #2  
+
+Option #1 ( Framework creates a Thread )
+---------
+Components register dump routine with debug framework using the API:  
+`Debugframework::linkWithFramework(std::string &componentName, const DumpInfoFunc funcPtr);`  
+
+Framework does the following:  
+- Update framework internal data structure to save the registration information of components.  
+- Create a `swss::Selectable` object.  
+- Create a subscriber to a table/channel in APPL_DB in RedisDB with a unique channel name for receiving triggers.  
+- Create a thread and wait on events/triggers.  
+- Invoke component specific dump routine in this thread context on receipt of a trigger.  
+- Handle the configured framework action within SWSS_DEBUG_PRINT. For example: Redirect the output to syslog or component specific log file.  
+- Update Redis DB to indicate callback action completion. 
+- Post-processing of information based on the configured framework action. For example: bundle the information and upload the information to a server.
+
+Option #2 ( Framework does not create a Thread )
+---------
+Components register dump routine with debug framework using the API:  
+`Debugframework::linkWithFrameworkNoThread(std::string &componentName);`  
+Framework will limit the job to the following:  
+- Update framework internal data structure to save the registration information of components.  
+- Handle the configured framework action within SWSS_DEBUG_PRINT. For example: Redirect the output to syslog or component specific log file.
+- Post-processing of information based on the configured framework action. For example: bundle the information and upload the information to a server.
+
+The below activity is delegated to the component
+- Optional creation of a swss::Selectable object. 
+- Create a subscriber instance to a table/channel in APPL_DB in RedisDB for receiving triggers.
+- Optionally create a thread to wake on triggers.
+- Invoke component specific dump routine on receipt of a trigger.
+- Update Redis DB to indicate callback action completion.
+
+
+**Implementation brief**  
+Framework will be implemented in C++ as a class in SWSS namespace.  

```

Comment:
how to open source project which are not written in C++, for example dhcp relay which is written in C?

Reply:
Good question.  For python programs SWIG will help us.  Check pyext folder for interface file definition. 

For C,  yes we need to write wrapper functions.